### PR TITLE
Allow `ChainEndpoint` implementations to fetch any types of clients and consensus states

### PR DESCRIPTION
**Pull request by romac**
Closes: #1481 

## TODO

- [x] Test that client upgrades still work

______

For contributor use:

- [x] Added a changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] If applicable: Unit tests written, added test to CI.
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: modules/src/core/ics02_client/client_state.rs [hunk 1]
```diff
@@ -40,6 +40,15 @@ pub trait ClientState: Clone + core::fmt::Debug + Send + Sync {
     /// Frozen height of the client
     fn frozen_height(&self) -> Option<Height>;
 
+    /// The client's unbonding period
+    fn unbonding_period(&self) -> Duration;
```
Comment (hu55a1n1) [inline modules/src/core/ics02_client/client_state.rs:44]: I think `unbonding_period` is Tendermint (or more generally PoS) specific so should not be a part of this trait.

File: modules/src/core/ics02_client/client_state.rs [hunk 2]
```diff
@@ -40,6 +40,15 @@ pub trait ClientState: Clone + core::fmt::Debug + Send + Sync {
...
+
+    /// Helper function to verify the upgrade client procedure.
+    /// Resets all fields except the blockchain-specific ones,
+    /// and updates the given fields.
+    fn upgrade(self, upgrade_height: Height, unbonding_period: Duration, chain_id: ChainId)
```
Comment (hu55a1n1) [inline modules/src/core/ics02_client/client_state.rs:49]: Same comment here regarding `unbonding_period` being specific to Tendermint blockchains. I think we might need a more generic upgrade params abstraction (maybe assoc type?).

**Review by hu55a1n1 — COMMENTED** [review]
Thanks, @romac! Nicely done! 👌 Just a couple of comments related to the separation of Tendermint specific code. I am trying to test the upgrade procedure manually and will follow up with the results soon.

File: modules/src/core/ics02_client/client_state.rs (hunk 1 above)
Comment (lightyear15) [reply modules/src/core/ics02_client/client_state.rs:44]: agreed
I am also concerned about ``AnyClientState::trust_threshold`` for the same reason :thinking:

File: relayer/src/chain.rs [hunk 3]
```diff
@@ -165,7 +165,7 @@ pub trait ChainEndpoint: Sized {
         &self,
         client_id: &ClientId,
         height: ICSHeight,
-    ) -> Result<Self::ClientState, Error>;
+    ) -> Result<AnyClientState, Error>;
```
Comment (lightyear15) [inline relayer/src/chain.rs:168]: I think the same change should be made for ``query_upgraded_client_state``

File: modules/src/clients/ics07_tendermint/consensus_state.rs [hunk 4]
```diff
@@ -44,7 +44,7 @@ impl crate::core::ics02_client::client_consensus::ConsensusState for ConsensusSt
     }
 
     fn validate_basic(&self) -> Result<(), Infallible> {
-        unimplemented!()
+        Ok(())
```
[the remark this answers is not in the document]
Comment (romac) [reply modules/src/clients/ics07_tendermint/consensus_state.rs]: 21f39674

File: modules/src/core/ics02_client/client_state.rs (hunk 2 above)
Comment (romac) [reply modules/src/core/ics02_client/client_state.rs]: Fixed in https://github.com/informalsystems/ibc-rs/pull/1625/commits/4b0a78dd082066a18cd11514af8e8b85ec5526f6

File: modules/src/core/ics02_client/client_state.rs (hunk 1 above)
Comment (romac) [reply modules/src/core/ics02_client/client_state.rs]: Fixed in https://github.com/informalsystems/ibc-rs/pull/1625/commits/4b0a78dd082066a18cd11514af8e8b85ec5526f6

File: relayer/src/chain.rs (hunk 3 above)
Comment (romac) [reply relayer/src/chain.rs:172]: https://github.com/informalsystems/ibc-rs/pull/1625/commits/9278d0460b926be2c640307302bb1d4ce8932f25

**Review by hu55a1n1 — APPROVED** [review]
❤️

**Pull request merged**